Return the correct OSVersion on OSX like systems.

eerhardt opened the pull request:
Fix #34977

Here's a [link to the underlying API](https://developer.apple.com/documentation/foundation/nsprocessinfo/1410906-operatingsystemversion?language=objc) being invoked. I tested this code both on a .NET 5 console app on macOS `10.15.4` and in an iOS simulator running iOS `13.4.1`. The method returned the respective version for each. I didn't test tvOS though - I don't have a device.

eerhardt committed:
Return the correct OSVersion on OSX like systems.

Fix #34977

am11 commented on src/libraries/System.Private.CoreLib/src/System/Environment.Linux.cs:
```diff
@@ -0,0 +1,67 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
For other Unix OS like {Free,Net}BSD default support, can this be renamed to `Environment.Unix.cs` and conditionally include `Environment.OSX.cs` in project file for macOS?

jkotas replied:
Maybe:

`Environment.OSVersion.Unix.cs`
`Environment.OSVersion.OSX.cs`

eerhardt replied:
I can add `OSVersion` to the file name. For OSX and `Unix vs. `Linux`, I was following the same pattern as FileStream:

https://github.com/dotnet/runtime/blob/16b7db27c98dce5b4d15894db13acd629b8fea62/src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems#L1759-L1760

Do you think that should be changed as well?

jkotas replied:
> Do you think that should be changed as well?

It would be nice.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Environment.OSX.cs:
```diff
@@ -0,0 +1,49 @@
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_msgSend(IntPtr basePtr, IntPtr selector);
+        [DllImport("libobjc.dylib", EntryPoint = "objc_msgSend_stret")]
+        private static extern NSOperatingSystemVersion get_operatingSystemVersion(IntPtr basePtr, IntPtr selector);
```
Why does this one have a different name than the entry point?

eerhardt replied:
The way `objc_msgSend_stret` works (from my understanding of reading docs and examples) is that the arguments and return values are mutable depending on the function's signature. The way I've seen C code calling it is by defining a function pointer with the right signature and casting `objc_msgSend_stret` to it.

I named the function `get_operatingSystemVersion` since the return value is specific to OS Version.

See https://blog.lazerwalker.com/objective-c,/code/2013/10/12/the-objective-c-runtime-and-objc-msgsend-stret.html

> The solution? You need to cast the objc_msgSend_stret function itself into a function whose return type is the struct you’re expecting. Here is a complete, functioning version of our example:

```
UIView *view = [[UIView alloc] initWithFrame:CGRectZero];

CGRect (*sendRectFn)(id receiver, SEL operation);
sendRectFn = (CGRect(*)(id, SEL))objc_msgSend_stret;
CGRect frame = sendRectFn(view, @selector(frame));
```

stephentoub commented on src/libraries/System.Runtime.Extensions/tests/System/EnvironmentTests.cs:
```diff
@@ -154,6 +154,18 @@ public void OSVersion_ParseVersion(string input, int major, int minor, int build
+            // verify that the Environment.OSVersion.Version matches the current RID
+            Assert.Contains(version.ToString(2), RuntimeInformation.RuntimeIdentifier);
+
+            Assert.Equal(-1, version.Revision); // Revision is never set on OSX
```
It looks like we're testing the 1st, 2nd, and 4th parts. Anything we can assert about the 3rd?

eerhardt replied:
When I first thought about it nothing came to mind. Thinking a little again, the one thing that does come to mind is that it is non-negative. I wouldn't think OSX would ever use a negative patch number.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Environment.OSX.cs:
```diff
@@ -0,0 +1,49 @@
+            public nint patchVersion;
+        }
+
+        [DllImport("libobjc.dylib")]
```
Is this is the right path to use?

The non-absolute paths tend to break on OSX when people happen to run with custom `DYLD_FALLBACK_LIBRARY_PATH`. See https://github.com/dotnet/runtime/issues/24095 for example

eerhardt replied:
This is what we were using for System.Drawing:

https://github.com/dotnet/runtime/blob/f5ecd6bc936cb9502d5faf8e156d00cf526c9b76/src/libraries/System.Drawing.Common/src/System/Drawing/macFunctions.cs#L211-L222

jkotas replied:
Does it work when you do `export DYLD_FALLBACK_LIBRARY_PATH=/nothing` ?

rolfbjarne replied:
We (the Xamarin.iOS/Mac team) always use full paths, because we've run into cases where it doesn't work.

I don't remember anymore exactly what happened, but we still have this in our code base: https://github.com/xamarin/xamarin-macios/blob/2926083d205fa47c402fa19397d2444d66dfaa2a/src/ObjCRuntime/Dlfcn.cs#L114-L126

It could have been something Apple broke in a beta (that code was originally committed on June 10, 2015, in the middle of WWDC/just after Apple released the first iOS 9 beta), and then fixed later on, but since then we've always used full paths, just in case Apple does something strange again.

stephentoub replied:
For reference, this is what Interop.Libraries.cs currently contains:
https://github.com/dotnet/runtime/blob/3b68eb3b1edc5a0ed0f4758668937fe4de17c431/src/libraries/Common/src/Interop/OSX/Interop.Libraries.cs#L5-L18

eerhardt replied:
I've added the full path: `/usr/lib/libobjc.dylib`

marek-safar commented on src/libraries/System.Private.CoreLib/src/System/Environment.OSX.cs:
```diff
@@ -0,0 +1,49 @@
+        }
+
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_getClass(string className);
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr sel_getUid(string selector);
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_msgSend(IntPtr basePtr, IntPtr selector);
```
Why are these declarations not in Common interop?

eerhardt replied:
No reason. I've moved them there.

rolfbjarne commented on src/libraries/System.Private.CoreLib/src/System/Environment.OSX.cs:
```diff
@@ -0,0 +1,49 @@
+        private static extern IntPtr sel_getUid(string selector);
+        [DllImport("libobjc.dylib")]
+        private static extern IntPtr objc_msgSend(IntPtr basePtr, IntPtr selector);
+        [DllImport("libobjc.dylib", EntryPoint = "objc_msgSend_stret")]
```
There's no `objc_msgSend_stret` on ARM64, you must call `objc_msgSend` instead.

eerhardt replied:
Nice call on this! I only tested the iOS simulator, not an actual device. Fixed.

eerhardt committed:
Rename files to not have Linux when they aren't Linux specific.

eerhardt committed:
Assert OSVersion Build number is valid.

eerhardt committed:
PR feedback

Move interop code to Common\src\Interop.
Specify the full path to libobjc.dylib.
Fix objc_msgSend_stret for ARM64.

jkotas commented on src/libraries/Common/src/Interop/OSX/Interop.Libraries.cs:
```diff
@@ -9,13 +9,13 @@ internal static partial class Libraries
         internal const string CoreServicesLibrary = "/System/Library/Frameworks/CoreServices.framework/CoreServices";
         internal const string CFNetworkLibrary = "/System/Library/Frameworks/CFNetwork.framework/CFNetwork";
+        internal const string libobjc = "/usr/lib/libobjc.dylib";
         internal const string libproc = "libproc";
```
Nit: You can do the same thing for libproc and fix https://github.com/dotnet/runtime/issues/24095 while you are on it.

eerhardt replied:
Good point. Might as well while I'm in here.

eerhardt committed:
Specify full path to libproc.dylib.

Fix #24095
